Clean up error messages for address of to non-function-pointer types.

333fred opened the pull request:
* Correctly error on direct cast to non-func ptr type
* Use better error message test.

Fixes https://github.com/dotnet/roslyn/issues/44489.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs for review of this small PR.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -2352,6 +2352,11 @@ private BoundExpression BindCastCore(ExpressionSyntax node, BoundExpression oper
                     }
+                case BoundKind.UnconvertedAddressOfOperator:
+                    {
+                        diagnostics.Add(ErrorCode.ERR_AddressOfToNonFunctionPointer , syntax.Location, ((BoundUnconvertedAddressOfOperator)operand).Operand.Name, targetType);
```
>ERR_AddressOfToNonFunctionPointer [](start = 50, length = 33)

How do we know the target type is not a function pointer or delegate? Consider adding explicit check.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -2138,12 +2138,12 @@ internal BoundExpression GenerateConversionForAssignment(TypeSymbol targetType,
                 case BoundKind.UnconvertedAddressOfOperator:
                     {
-                        reportMethodGroupErrors(((BoundUnconvertedAddressOfOperator)operand).Operand);
+                        reportMethodGroupErrors(((BoundUnconvertedAddressOfOperator)operand).Operand, fromFunctionPtr: true);
```
>fromFunctionPtr: true [](start = 102, length = 21)

This feels somewhat misleading, an "address of" operator is not a pointer.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3656,24 +3656,132 @@ .maxstack  1
+    unsafe void M()
+    {
+        // This actually gets bound as a binary expression: (Action) & M
+        Action ptr1 = (Action)&M;
```
>(Action)&M [](start = 22, length = 10)

Consider also testing ```(Action)(&M)```

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3656,24 +3656,132 @@ .maxstack  1
+    unsafe void M()
+    {
+        // This actually gets bound as a binary expression: (C) & M
+        C ptr1 = (C)&M;
```
>(C)&M [](start = 17, length = 5)

Same suggestion.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3656,24 +3656,132 @@ .maxstack  1
         }
 
+
         [Fact]
```
>[Fact] [](start = 8, length = 6)

It looks like an extra empty line is added above

AlekseyTs commented:
It looks like there are legitimate test failures.

AlekseyTs commented:
Done with review pass (iteration 2)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [635197431](https://github.com/dotnet/roslyn/pull/44613#issuecomment-635197431) [](ancestors = 635197431)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3656,22 +3656,182 @@ .maxstack  1
+            FunctionPointerUtilities.VerifyFunctionPointerSemanticInfo(model, ((CastExpressionSyntax)decls[1].Initializer!.Value).Expression,
+                expectedSyntax: "(&M)",
+                expectedType: null,
+                expectedConvertedType: null,
```
>null [](start = 39, length = 4)

The difference with the check below feels unexpected. Why does it matter whether there was an explicit cast in code?

333fred replied:
This mirrors how explicit conversions work in general: a `TypeInfo` does not contain information about explicit conversions. This mirrors how `(Action)M` would perform, though I do actually need to change the conversion to `Identity`, not `NoConversion`.

---
In reply to: [432042922](https://github.com/dotnet/roslyn/pull/44613#discussion_r432042922) [](ancestors = 432042922)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3656,22 +3656,182 @@ .maxstack  1
+            FunctionPointerUtilities.VerifyFunctionPointerSemanticInfo(model, ((CastExpressionSyntax)decls[1].Initializer!.Value).Expression,
+                expectedSyntax: "(&M)",
+                expectedType: null,
+                expectedConvertedType: null,
```
>null [](start = 39, length = 4)

Same coment here

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -2115,6 +2115,14 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
+                    type = null;
+                    convertedType = null;
+                    convertedNullability = nullability;
+                    conversion = Conversion.NoConversion;
```
>conversion = Conversion.NoConversion; [](start = 20, length = 37)

Are we testing this aspect?

333fred replied:
It's verified by the utility, but I need to update both this and it to have an `Identity` conversion like `(Action)M` would have.

---
In reply to: [432045889](https://github.com/dotnet/roslyn/pull/44613#discussion_r432045889) [](ancestors = 432045889)

AlekseyTs commented:
Done with review pass (iteration 3)

333fred commented:
@AlekseyTs this is ready for another look.

cston commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -1806,6 +1806,8 @@ internal enum ErrorCode
         ERR_DupReturnTypeMod = 8798,
         ERR_AddressOfMethodGroupInExpressionTree = 8799,
+        ERR_CannotConvertAddressOfToDelegate = 8800,
+        ERR_AddressOfToNonFunctionPointer = 8801,
```
The error codes appear to be used already. (Most are used in master and 8801 is used in [https://github.com/dotnet/roslyn/pull/44636](https://github.com/dotnet/roslyn/pull/44636).)

333fred replied:
I'm already going to have to resolve conflicts here, one or two more isn't particularly more difficult.

---
In reply to: [432171912](https://github.com/dotnet/roslyn/pull/44613#discussion_r432171912) [](ancestors = 432171912)

cston commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6160,4 +6160,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>'RefKind.Out' is not a valid ref kind for a return type.</value>
   </data>
+  <data name="ERR_CannotConvertAddressOfToDelegate" xml:space="preserve">
+    <value>Cannot convert a &amp; method group '{0}' to delegate type '{0}'.</value>
```
>convert a &amp; method group [](start = 18, length = 28)

This message uses "convert a & method group" while the message below uses "convert & method group". We should be consistent. Consider dropping the "a" since it's not clear whether it should be "a" or "an".

333fred replied:
I'll correct this in my IDE pr I have out now.

---
In reply to: [432172457](https://github.com/dotnet/roslyn/pull/44613#discussion_r432172457) [](ancestors = 432172457)
